src/librustc_mir/transform/const_prop.rs

```diff
@@ -969,13 +970,40 @@ impl<'mir, 'tcx> MutVisitor<'tcx> for ConstPropagator<'mir, 'tcx> {
             | TerminatorKind::GeneratorDrop
             | TerminatorKind::FalseEdges { .. }
             | TerminatorKind::FalseUnwind { .. } => {}
-            //FIXME(wesleywiser) Call does have Operands that could be const-propagated
-            TerminatorKind::Call { .. } => {}
+            TerminatorKind::Call { ref mut args, .. } => {
+                let mir_opt_level = self.tcx.sess.opts.debugging_opts.mir_opt_level;
+                // Constant Propagation into function call arguments is gated
+                // under mir-opt-level 2, because LLVM codegen gives performance
+                // regressions with it.
+                if mir_opt_level >= 2 {
+                    // Every argument in our function calls can be const propagated.
+                    // We try to, before exiting.
+                    for opr in args {
+                        if let Some(l) = opr.place().and_then(|p| p.as_local()) {
+                            if let Some(value) = self.get_const(l) {
```

Comment:
I think on your original PR you asked why this approach was ok? The reasoning is that until very recently, we always spilled function call arguments to temporaries (basically, `add(1u8, 2u8)` would become `_tmp0 = 1u8; _tmp1 = 2u8; add(_tmp0, _tmp1`). The reasons for doing that aren't terribly important. 

The only caveat here is an unstable feature called `unsized_locals`. You can read more about what changed in #71170 but the TLDR is that in a very niche situation, we don't follow this exact pattern. I think this is unimportant for const-prop because 1) this is a very specific situation and 99.999% of Rust code in the wild is not going to do this and 2) I suspect const-prop can't handle derefs to unsized locals anyway so it would have given up before even reaching this line of code.

Reply:
> The reasoning is that until very recently, we always spilled function call arguments to temporaries (basically, `add(1u8, 2u8)` would become `_tmp0 = 1u8; _tmp1 = 2u8; add(_tmp0, _tmp1`).

When did it change? I ask because as you can see, this PR only [changes the `scalar_literal_propagation` test](https://github.com/rust-lang/rust/pull/71697/files#diff-af037b01d0beedd3aaf474787962616b), and not [the `tuple_literal_propagation` test](https://github.com/rust-lang/rust/blob/master/src/test/mir-opt/const_prop/tuple_literal_propagation/rustc.main.ConstProp.diff).

If we still spilled function call arguments to temporaries, tuples would be const propagated here. I assume we're not spilling them anymore, because the tuple test shows no change. 
Is that right? @wesleywiser

Reply:
> When did it change? I ask because as you can see, this PR only changes the scalar_literal_propagation test, and not the tuple_literal_propagation test.

This is related to the comment thread below. In that test, the argument is a `ScalarPair(Scalar(1u8), Scalar(2u8))` but on line 986 of const_prop.rs, we only handle the case of `Scalar(_)` which is why that test isn't updated in this PR. 

----

Anyway, that's not really related to what I was saying in this thread. We still always spill arguments to temporaries except for unsized locals passed as arguments. This behavior is not changed by your PR (spilling the arguments happens during MIR constructions long before const prop runs). Take a look at https://github.com/rust-lang/rust/pull/71170 for more info on the unsized locals thing.

Reply:
Ahh, okay okay, I think I understand now. You don't mean to say that `ScalarPairs` are spilled and converted into two arguments of `Scalar` kind. You meant that for example a `ScalarPair` that's used as argument for a function is actually used through a `Temporary` - not that its components are converted to individual `Operands`. Right?

Reply:
Yes, you got it!

Reply:
Awesome. Okay, right after we find out about that ICE in which we were pinged (and oops, I hope it wasn't our PR ^), which I'm trying to use to learn rust-bisect, I'll write this down in comments :)

Reply:
Okay, after an exercise break and re-reading all of our discussion regarding this snippet, I don't think it's as important as I thought it was, to document why it's the way it is.

I originally thought that we were somehow discriminating against `ScalarPairs`, or that's what I remember at least. Now that I read you better, I understand that this code is special because of the `Temporary` spillover, and that it has nothing much to do in how we want to treat `ScalarPairs`.

If our plan is to just implement the propagation as it is right now (only propagating `Scalars`), and leaving propagation of `ScalarPairs` to the eventual generalization of our constant propagation into `Operands` code, I'm more than happy to follow that plan.

Reply:
So, I think I now understand why this snippet, which is simpler than the original one I made, is perfectly fine.

In this code, we don't really handle the cases where the `Operand` is of variant `Operand::Constant`, because then the `place()` function returns `None`. However, since we do the `Temporary` spillover, any `Place` that was used by `Constant` or `Move` is copied into an `Operand::Copy` variant. Which allows us to do `const-prop` over it. Is this correct? @wesleywiser

Reply:
Yep, that sounds exactly right!